### PR TITLE
Ntfs parser bug fix

### PR DESCRIPTION
**Pull request by danilbaz**
- Correct a bug causing a memory saturation of the remote machine
- Decrease the number of call of railgun
- Add debug output

### CODE REVIEW

File: lib/rex/parser/fs/ntfs.rb [hunk 1]
```diff
@@ -181,7 +181,8 @@ def cluster_from_attribute_non_resident(attribute, cluster_num = 0, size_max = (
 
             data = ''
             while data.length < size_wanted
-              data << @file_handler.read(size_wanted - data.length)
+              # use an arbitrary block size to avoid target memory consomation
```
Comment (zeroSteiner) [inline lib/rex/parser/fs/ntfs.rb]: target memory consumption?

**Comment by zeroSteiner** [comment]
Do you have steps that could be followed to reproduce the bug that this PR addresses?

File: lib/rex/parser/fs/ntfs.rb [hunk 2]
```diff
@@ -213,10 +217,11 @@ def mft_record_attribute(mft_record)
             res[attribute_identifier] = mft_record[curs + content_offset, content_size]
           else
             # non resident
-            if attribute_identifier == DATA_ATTRIBUTE_ID
-              res[attribute_identifier] = mft_record[curs, attribute_size]
-            else
+            if attribute_identifier == INDEX_ALLOCATION_ID or 
+              (not lazy and attribute_identifier != DATA_ATTRIBUTE_ID)
```
Comment (zeroSteiner) [inline lib/rex/parser/fs/ntfs.rb]: ```!lazy``` would be the preferred logical operator over ```not lazy```.

Comment (zeroSteiner) [inline lib/rex/parser/fs/ntfs.rb]: I don't think ```and attribute_identifier != DATA_ATTRIBUTE_ID``` is necessary here. Wouldn't ```if attribute_identifier == INDEX_ALLOCATION_ID or !lazy``` suffice?

**Comment by danilbaz** [comment]
@zeroSteiner  The bug can be reproduced by trying this module on a large file on a machine with few memory (for example C:\pagefile.sys)

File: lib/rex/parser/fs/ntfs.rb (hunk 2 above)
Comment (danilbaz) [inline lib/rex/parser/fs/ntfs.rb]: @zeroSteiner the attribute DATA, must not be gathered even in not lazy mode (too huge).
the attribute index_allocation must always be gathered, it's necessary to parse directory.
Otherway, the attribute must be gathered only in not lazy mode.
I think that's what i implement.

File: lib/rex/parser/fs/ntfs.rb [hunk 3]
```diff
@@ -181,7 +181,8 @@ def cluster_from_attribute_non_resident(attribute, cluster_num = 0, size_max = (
 
             data = ''
             while data.length < size_wanted
-              data << @file_handler.read(size_wanted - data.length)
+              # use an arbitrary block size to avoid target memory consumption
+              data << @file_handler.read([size_wanted - data.length, 2**26].min)
```
Comment (danilbaz) [inline lib/rex/parser/fs/ntfs.rb]: This arbitrary block size can be optimized.
For example 2**21 allow a 50Mb/s transfert on a network that support it.
More or less slow down the transfer.
I don't know what compromise choose.

File: lib/rex/parser/fs/ntfs.rb [hunk 4]
```diff
@@ -181,7 +181,8 @@ def cluster_from_attribute_non_resident(attribute, cluster_num = 0, size_max = (
...
             end
             attribut << data
```
Comment (todb-r7) [inline lib/rex/parser/fs/ntfs.rb]: Not your problem with this PR, but I just spent 100 seconds trying to figure out if `attribut` was a typo or a real variable name. Gleesh.

File: lib/rex/parser/fs/ntfs.rb (hunk 3 above)
Comment (hmoore-r7) [inline lib/rex/parser/fs/ntfs.rb]: This 64Mb block size seems reasonable, although smaller may be nicer when using Metasploit on machines with lower amounts of memory. Maybe 4Mb (2**22)?

**Pull request merged**